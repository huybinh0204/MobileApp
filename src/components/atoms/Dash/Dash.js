import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes } from 'react-native';
import { defaultStyles, getDashStyle, isStyleRow } from './Dash.styles';
import { useComponentSize } from '_hooks';
import { COLORS } from '_constants';

const Dash = (props) => {
  const [size, layout] = useComponentSize();
  const isRow = isStyleRow(props.style);
  const length = isRow ? size.width : size.height;
  const n = Math.ceil(length / (props.dashGap + props.dashLength));
  const calculatedDashStyles = getDashStyle(props);

  let dash = [];
  for (let i = 0; i < n; i++) {
    dash.push(<View key={i} style={[calculatedDashStyles, props.dashStyle]} />);
  }

  return (
    <View onLayout={layout} style={[props.style, isRow ? defaultStyles.row : defaultStyles.column]}>
      {dash}
    </View>
  );
};

Dash.propTypes = {
  style: ViewPropTypes.style,
  dashGap: PropTypes.number.isRequired,
  dashLength: PropTypes.number.isRequired,
  dashThickness: PropTypes.number.isRequired,
  dashColor: PropTypes.string,
  dashStyle: ViewPropTypes.style,
};

Dash.defaultProps = {
  dashGap: 2,
  dashLength: 4,
  dashThickness: 2,
  dashColor: COLORS.black,
};

export default Dash;
