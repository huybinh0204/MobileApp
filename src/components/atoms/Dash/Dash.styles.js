import { css } from '@emotion/native';
import { StyleSheet } from 'react-native';

export const isStyleRow = (style) => {
  const flatStyle = StyleSheet.flatten(style || {});

  return flatStyle.flexDirection !== 'column';
};

const getDashStyleId = ({ dashGap, dashLength, dashThickness, dashColor }, isRow) =>
  `${dashGap}-${dashLength}-${dashThickness}-${dashColor}-${isRow ? 'row' : 'column'}`;

const createDashStyle = ({ dashGap, dashLength, dashThickness, dashColor }, isRow) => {
  return css({
    width: isRow ? dashLength : dashThickness,
    height: isRow ? dashThickness : dashLength,
    marginRight: isRow ? dashGap : 0,
    marginBottom: isRow ? 0 : dashGap,
    backgroundColor: dashColor,
  });
};

export const defaultStyles = {
  row: css({ flexDirection: 'row' }),
  column: css({ flexDirection: 'column' }),
};

let stylesStore = {};
export const getDashStyle = (props) => {
  const isRow = isStyleRow(props.style);
  const id = getDashStyleId(props, isRow);

  if (!stylesStore[id]) {
    stylesStore = {
      ...stylesStore,
      [id]: createDashStyle(props, isRow),
    };
  }

  return stylesStore[id];
};
