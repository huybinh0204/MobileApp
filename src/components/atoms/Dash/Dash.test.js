import React from 'react';
import { render } from '@testing-library/react-native';
import Dash from './Dash';

describe('Dash', () => {
  it('should match the snapshot', () => {
    const { toJSON } = render(
      <Dash
        style={{
          width: 1,
          height: 30,
        }}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
