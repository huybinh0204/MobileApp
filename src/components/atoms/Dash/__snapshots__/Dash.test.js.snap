// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dash should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "height": 30,
        "width": 1,
      },
      Object {
        "flexDirection": "row",
      },
    ]
  }
/>
`;
